In [3]:
import os
os.getcwd()

'/Users/eniola/Documents/Breakthrough tech/Swytch/Swytch-1A-do-ai-job-risk-indices-agree/notebooks'

In [4]:
import pandas as pd

df = pd.read_csv('../data/datasets/occ_level.csv') 

#Look into the first 20 rows of the dataset
df.head(20)


,O*NET-SOC Code,Title,dv_rating_alpha,dv_rating_beta,dv_rating_gamma,human_rating_alpha,human_rating_beta,human_rating_gamma
0,11-1011.00,Chief Executives,0.100000,0.460000,0.820000,0.180000,0.350000,0.520000
1,11-1011.03,Chief Sustainability Officers,0.166667,0.555556,0.944444,0.055556,0.388889,0.722222
2,11-1021.00,General and Operations Managers,0.000000,0.480769,0.961538,0.115385,0.384615,0.653846
3,11-1031.00,Legislators,0.033333,0.400000,0.766667,0.266667,0.516667,0.766667
4,11-2011.00,Advertising and Promotions Managers,0.000000,0.476744,0.953488,0.255814,0.546512,0.837209
5,11-2021.00,Marketing Managers,0.062500,0.500000,0.937500,0.218750,0.578125,0.937500
6,11-2022.00,Sales Managers,0.000000,0.450000,0.900000,0.100000,0.483333,0.866667
7,11-2032.00,Public Relations Managers,0.150000,0.550000,0.950000,0.250000,0.575000,0.900000
8,11-2033.00,Fundraising Managers,0.125000,0.562500,1.000000,0.250000,0.500000,0.750000
9,11-3012.00,Administrative Services Managers,0.086957,0.391304,0.695652,0.260870,0.500000,0.739130


In [5]:
#Data exploration to ensure we have no null values
print(df.isnull().sum())

df.info()

O*NET-SOC Code        0
Title                 0
dv_rating_alpha       0
dv_rating_beta        0
dv_rating_gamma       0
human_rating_alpha    0
human_rating_beta     0
human_rating_gamma    0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 923 entries, 0 to 922
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   O*NET-SOC Code      923 non-null    object 
 1   Title               923 non-null    object 
 2   dv_rating_alpha     923 non-null    float64
 3   dv_rating_beta      923 non-null    float64
 4   dv_rating_gamma     923 non-null    float64
 5   human_rating_alpha  923 non-null    float64
 6   human_rating_beta   923 non-null    float64
 7   human_rating_gamma  923 non-null    float64
dtypes: float64(6), object(2)
memory usage: 57.8+ KB


In [6]:
# Glance at O*NET-SOC Code that have values after the decimal point
onet = df['O*NET-SOC Code']

special_row = []
number_to = []

for num in onet:
    if int(num[8:10]) > 0:
        detail_row = df[df['O*NET-SOC Code'] == num]
        number_to.append(num)
        special_row.append(detail_row)

detail_rows = pd.concat(special_row)  

print("Number of rows with O*NET-SOC Code having values after the decimal point:", len(detail_rows))
detail_rows

Number of rows with O*NET-SOC Code having values after the decimal point: 149


,O*NET-SOC Code,Title,dv_rating_alpha,dv_rating_beta,dv_rating_gamma,human_rating_alpha,human_rating_beta,human_rating_gamma
1,11-1011.03,Chief Sustainability Officers,0.166667,0.555556,0.944444,0.055556,0.388889,0.722222
11,11-3013.01,Security Managers,0.071429,0.482143,0.892857,0.250000,0.535714,0.821429
14,11-3031.01,Treasurers and Controllers,0.000000,0.473684,0.947368,0.157895,0.407895,0.657895
15,11-3031.03,Investment Fund Managers,0.000000,0.500000,1.000000,0.171429,0.528571,0.885714
17,11-3051.01,Quality Control Systems Managers,0.075472,0.509434,0.943396,0.188679,0.471698,0.754717
...,...,...,...,...,...,...,...,...
867,51-9195.05,"Potters, Manufacturing",0.000000,0.083333,0.166667,0.000000,0.011905,0.023810
873,53-1042.01,Recycling Coordinators,0.137931,0.379310,0.620690,0.310345,0.465517,0.620690
904,53-6051.01,Aviation Inspectors,0.080000,0.340000,0.600000,0.120000,0.260000,0.400000
905,53-6051.07,"Transportation Vehicle, Equipment and Systems ...",0.208333,0.458333,0.708333,0.000000,0.125000,0.250000


In [7]:
#Get list of column names which we would operate on and convert to numeric values

list(df.columns)
operate_on = df.drop(columns=['O*NET-SOC Code', 'Title']).columns.tolist()
print("Columns to operate on:", operate_on)

df[operate_on] = df[operate_on].apply(pd.to_numeric, errors='coerce')
df.dtypes


Columns to operate on: ['dv_rating_alpha', 'dv_rating_beta', 'dv_rating_gamma', 'human_rating_alpha', 'human_rating_beta', 'human_rating_gamma']


O*NET-SOC Code         object
Title                  object
dv_rating_alpha       float64
dv_rating_beta        float64
dv_rating_gamma       float64
human_rating_alpha    float64
human_rating_beta     float64
human_rating_gamma    float64
dtype: object

In [8]:
# Create a title_lookup so we can pair the correct title and O*NET-SOC (Base) Code with the correct row in the dataframe
title_lookup = df[df['O*NET-SOC Code'].str.endswith('.00')][['O*NET-SOC Code', 'Title']]
title_lookup

,O*NET-SOC Code,Title
0,11-1011.00,Chief Executives
2,11-1021.00,General and Operations Managers
3,11-1031.00,Legislators
4,11-2011.00,Advertising and Promotions Managers
5,11-2021.00,Marketing Managers
...,...,...
918,53-7071.00,Gas Compressor and Gas Pumping Station Operators
919,53-7072.00,"Pump Operators, Except Wellhead Pumpers"
920,53-7073.00,Wellhead Pumpers
921,53-7081.00,Refuse and Recyclable Material Collectors


In [9]:
# Make sure soc_6digit columns exist --> removes the decimal point and everything after it, leaving only the 6-digit family code
df['soc_6digit'] = df['O*NET-SOC Code'].str.split('.').str[0]

# Group all rows (base + detail codes) by their 6-digit family, and calculate the mean of each column in operate_on
df_processed = df.groupby('soc_6digit')[operate_on].mean().reset_index()



# Add soc_6digit to title_lookup so it can be merged with df_processed
title_lookup['soc_6digit'] = df['soc_6digit']

# Merge titles into df_processed by soc_6digit
df_processed = df_processed.merge(title_lookup[['soc_6digit', 'Title']], on='soc_6digit', how='left')

# Move Title to right after soc_6digit
df_processed.insert(1, 'Title', df_processed.pop('Title'))

print(f"Original rows: {len(df)}, After rollup: {len(df_processed)}")
    

Original rows: 923, After rollup: 798


In [10]:
#Display the final dataset after rollup and merging titles
df_processed

,soc_6digit,Title,dv_rating_alpha,dv_rating_beta,dv_rating_gamma,human_rating_alpha,human_rating_beta,human_rating_gamma
0,11-1011,Chief Executives,0.133333,0.507778,0.882222,0.117778,0.369444,0.621111
1,11-1021,General and Operations Managers,0.000000,0.480769,0.961538,0.115385,0.384615,0.653846
2,11-1031,Legislators,0.033333,0.400000,0.766667,0.266667,0.516667,0.766667
3,11-2011,Advertising and Promotions Managers,0.000000,0.476744,0.953488,0.255814,0.546512,0.837209
4,11-2021,Marketing Managers,0.062500,0.500000,0.937500,0.218750,0.578125,0.937500
...,...,...,...,...,...,...,...,...
793,53-7071,Gas Compressor and Gas Pumping Station Operators,0.230769,0.269231,0.307692,0.000000,0.076923,0.153846
794,53-7072,"Pump Operators, Except Wellhead Pumpers",0.166667,0.250000,0.333333,0.000000,0.166667,0.333333
795,53-7073,Wellhead Pumpers,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
796,53-7081,Refuse and Recyclable Material Collectors,0.178571,0.196429,0.214286,0.107143,0.232143,0.357143


# Note: SOC 8-digit → 6-digit Rollup — Reduction Count Discrepancy

## Finding

The raw dataset contains **923** occupations at the 8-digit O*NET-SOC level. After
rolling up to 6-digit SOC codes, the final dataset contains **798** unique codes —
a reduction of **125** rows.

At first glance this doesn't match expectations: there are **149** detail-level
rows in the raw data (codes with a suffix other than `.00`). If every detail row
simply merged into an existing base (`.00`) row, the reduction should also be 149,
not 125.

## Why the numbers differ

The gap (149 − 125 = **24**) is explained by **24 six-digit SOC families that have
no `.00` base row at all** — they exist *only* as detail-level codes. For example,
`29-1299` and `19-3099` have multiple `.XX` detail codes but no general `.00` entry.

For these 24 families, the detail rows aren't "extra" rows stacked on top of a base
— they *are* the entire group. Collapsing them into one 6-digit row doesn't reduce
the row count the same way it does for a family that already had a base row, which
is why the total reduction comes out lower than a naive count of detail rows would
suggest.



In [11]:
#Verification of the rollup and merging of titles

df['suffix'] = df['O*NET-SOC Code'].str.split('.').str[1].astype(int)

group_sizes = df.groupby('soc_6digit').size()

has_base = set(df.loc[df['suffix'] == 0, 'soc_6digit'].unique())
all_groups = set(group_sizes.index)
no_base_groups = all_groups - has_base

detail_count = (df['suffix'] != 0).sum()
excess = (group_sizes - 1).sum()

print(f"Total rows: {len(df)}")
print(f"Detail rows (suffix != .00): {detail_count}")
print(f"Unique 6-digit groups: {len(group_sizes)}")
print(f"Reduction: {len(df) - len(group_sizes)}")
print(f"6-digit codes with NO .00 base row: {len(no_base_groups)}")
print(sorted(no_base_groups))

print("Reconciliation check — detail_count - excess should equal no_base count:",
      detail_count - excess, "vs", len(no_base_groups))

Total rows: 923
Detail rows (suffix != .00): 149
Unique 6-digit groups: 798
Reduction: 125
6-digit codes with NO .00 base row: 24
['11-9179', '11-9199', '13-1199', '13-2099', '15-1299', '15-2099', '17-2199', '17-3029', '19-1029', '19-2099', '19-3039', '19-3099', '19-4099', '25-2059', '29-1129', '29-1229', '29-1299', '29-2099', '29-9099', '31-9099', '33-9099', '47-4099', '49-9099', '51-8099']
Reconciliation check — detail_count - excess should equal no_base count: 24 vs 24


In [12]:
#Spot-check for rows that did not have a base row, and see if they were rolled up correctly
df[df['soc_6digit'] == '29-1299'][['O*NET-SOC Code', 'dv_rating_alpha']]
df_processed[df_processed['soc_6digit'] == '29-1299']

,soc_6digit,Title,dv_rating_alpha,dv_rating_beta,dv_rating_gamma,human_rating_alpha,human_rating_beta,human_rating_gamma
336,29-1299,NaN,0.196314,0.496394,0.796474,0.145032,0.315705,0.486378


## Status: Rollup Complete — Title Resolution Pending

Most of the rollup operation is now complete: the dataset has been successfully
collapsed from 923 occupations (8-digit O*NET-SOC) to 798 unique 6-digit SOC
codes, with the numeric exposure columns aggregated via mean.

**Remaining issue:** 24 of the 798 rows are missing a `Title`. This is because
these 6-digit SOC families have no `.00` base row to pull a title from — they
exist only as detail-level codes (see prior note on the reduction discrepancy).
Since a straightforward groupby has no title to inherit for these rows, they
need to be resolved separately, most likely by looking up the official 6-digit
SOC title from the BLS classification system rather than arbitrarily picking
one of the detail titles.

This will be addressed in the next step before the dataset is finalized.

In [13]:
no_base_groups

{'11-9179',
 '11-9199',
 '13-1199',
 '13-2099',
 '15-1299',
 '15-2099',
 '17-2199',
 '17-3029',
 '19-1029',
 '19-2099',
 '19-3039',
 '19-3099',
 '19-4099',
 '25-2059',
 '29-1129',
 '29-1229',
 '29-1299',
 '29-2099',
 '29-9099',
 '31-9099',
 '33-9099',
 '47-4099',
 '49-9099',
 '51-8099'}

In [14]:
# Get Titles for the 6-digit codes that did not have a base row from original dataset
title_lookup_no_base = df[df['O*NET-SOC Code'].str.split('.').str[0].isin(no_base_groups)][['soc_6digit', 'Title']]
title_lookup_no_base

,soc_6digit,Title
48,11-9179,Fitness and Wellness Coordinators
49,11-9179,Spa Managers
50,11-9199,Regulatory Affairs Managers
51,11-9199,Compliance Managers
52,11-9199,Loss Prevention Managers
...,...,...
486,31-9099,Endoscopy Technicians
512,33-9099,Retail Loss Prevention Specialists
701,47-4099,Weatherization Installers and Technicians
763,49-9099,Geothermal Technicians


In [21]:
group_soc = title_lookup_no_base.groupby('soc_6digit')['Title'].apply(list)
#for better visibility
for soc_code, titles in group_soc.items():
    print(f"\n{soc_code}:")
    for t in titles:
        print(f"  - {t}")


11-9179:
  - Fitness and Wellness Coordinators
  - Spa Managers

11-9199:
  - Regulatory Affairs Managers
  - Compliance Managers
  - Loss Prevention Managers
  - Wind Energy Operations Managers
  - Wind Energy Development Managers
  - Brownfield Redevelopment Specialists and Site Managers

13-1199:
  - Business Continuity Planners
  - Sustainability Specialists
  - Online Merchants
  - Security Management Specialists

13-2099:
  - Financial Quantitative Analysts
  - Fraud Examiners, Investigators and Analysts

15-1299:
  - Web Administrators
  - Geographic Information Systems Technologists and Technicians
  - Document Management Specialists
  - Penetration Testers
  - Information Security Engineers
  - Digital Forensics Analysts
  - Blockchain Engineers
  - Computer Systems Engineers/Architects
  - Information Technology Project Managers

15-2099:
  - Bioinformatics Technicians

17-2199:
  - Energy Engineers, Except Wind and Solar
  - Mechatronics Engineers
  - Microsystems Engineers

In [22]:
#Final Dataframe to csv
df_processed.to_csv('../data/datasets/eloundou_6digit.csv', index=False)

## Status: SOC 8-digit → 6-digit Rollup Complete
 
This notebook rolls up 8-digit O*NET-SOC exposure data to 6-digit SOC codes.
 
- Starting point: **923** occupations at the 8-digit O*NET-SOC level (`occ_level.csv`)
- Rows were grouped by their 6-digit SOC prefix, and all numeric exposure columns
  were aggregated using the **mean**
- Titles were pulled from each family's `.00` base code where one exists
- The rollup produced **798** unique 6-digit SOC codes
This reduction (923 → 798) was verified against the underlying detail-row counts,
including reconciling **24** SOC families that exist *only* as detail-level codes
(no `.00` base row).
 
**Known gap:** those 24 rows currently have no `Title` populated. This is a known,
documented limitation — not an error — pending a future crosswalk against the
official BLS SOC structure. See the standalone methods note for rationale.
 

# Methods Note: SOC 8-digit → 6-digit Rollup

## Rollup Rule

Exposure scores were rolled up from the 8-digit O*NET-SOC level to the 6-digit
SOC level using a **simple (unweighted) mean** across all detail rows within
each 6-digit family.

## Rationale

Mean was chosen as a fast, defensible starting point to keep the pipeline
moving — not as a deliberated final answer. Alternatives (e.g. employment-weighted
mean using OES employment counts, or min/max bounding for conservative vs.
liberal exposure estimates) were **not evaluated in depth** at this stage.
That comparison is better suited to a team-level discussion during the
crosswalk stage, since the choice affects every downstream index comparison.
The current mean-based approach is easy to swap out later if the team decides
on a different weighting scheme.

## Known Limitation: Missing Titles

24 of the 798 six-digit SOC codes have no `Title` populated. This is because
these families have no `.00` base row to pull a title from — they exist only
as detail-level O*NET-SOC codes.

This has been **left as a documented gap rather than resolved ad hoc**:
- Resolving it properly requires a crosswalk against the official BLS SOC
  structure/title list, which is outside this task's scope.
- Manually assigning titles from memory or guesswork risked introducing
  incorrect labels and could conflict with the separate crosswalk effort
  another team member may own.

## Verification

The 923 → 798 row reduction was checked against the underlying data:
149 detail-level rows (non-`.00` suffix) exist in the raw dataset, but the
net reduction is 125, not 149. This gap (24) is fully explained by the 24
SOC families noted above that have no base row — their detail rows collapse
into a single group without producing the same row reduction as families
that already had a base row. This verification is considered solid and
stated here as fact, not an open question.